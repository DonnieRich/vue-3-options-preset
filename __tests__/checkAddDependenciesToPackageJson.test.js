const { vol } = require("memfs");
jest.mock("fs/promises");

const { addDependenciesToPackageJson } = require("../src/bin/addDependenciesToPackageJson");
const { jsonOperations } = require("../src/bin/jsonOperations");

const BASE_FILE = '/package.json';
const BASE_STUB_FILE = '/node_modules/vue-3-options-preset/src/stubs/package.json';

// preparing the mock modules
let mockGetJsonDataFromFile;
let mockMergeJsonObjects;
let mockWriteJsonDataToFile;

describe(addDependenciesToPackageJson, () => {
    beforeEach(() => {
        vol.reset();
        mockGetJsonDataFromFile = jest.spyOn(jsonOperations, "getJsonDataFromFile");
        mockMergeJsonObjects = jest.spyOn(jsonOperations, "mergeJsonObjects");
        mockWriteJsonDataToFile = jest.spyOn(jsonOperations, "writeJsonDataToFile");
    });

    afterEach(() => {
        mockGetJsonDataFromFile.mockRestore();
        mockMergeJsonObjects.mockRestore();
        mockWriteJsonDataToFile.mockRestore();
    });

    const stubJson = {
        dependencies: {
            "@popperjs/core": "^2.11.6",
            bootstrap: "^5.2.3"
        },
        devDependencies: {
            sass: "^1.57.1"
        }
    };

    const baseJson = {
        license: "MIT",
        dependencies: {
            shelljs: "^0.8.5"
        },
        devDependencies: {
            jest: "^29.3.1",
            memfs: "^3.4.12"
        }
    };

    const updatedJsonNoBootstrap = {
        license: "MIT",
        dependencies: {
            shelljs: "^0.8.5"
        },
        devDependencies: {
            sass: "^1.57.1",
            jest: "^29.3.1",
            memfs: "^3.4.12"
        }
    };

    const updatedJsonBootstrap = {
        license: "MIT",
        dependencies: {
            shelljs: "^0.8.5",
            "@popperjs/core": "^2.11.6",
            bootstrap: "^5.2.3"
        },
        devDependencies: {
            sass: "^1.57.1",
            jest: "^29.3.1",
            memfs: "^3.4.12"
        }
    };

    it("It should pass if the file package.json has been updated with the merged devDependencies", async () => {

        vol.fromJSON(
            {
                [BASE_FILE]: JSON.stringify(baseJson),
                [BASE_STUB_FILE]: JSON.stringify(stubJson)
            },
            '/'
        );

        await addDependenciesToPackageJson(BASE_STUB_FILE, BASE_FILE, false);

        // check calls to getJsonDataFromFile
        expect(mockGetJsonDataFromFile).toHaveBeenNthCalledWith(1, BASE_STUB_FILE);
        expect(mockGetJsonDataFromFile).toHaveBeenNthCalledWith(2, BASE_FILE);

        // check calls to mergeJsonObjects
        expect(mockMergeJsonObjects).toHaveBeenNthCalledWith(1, stubJson.devDependencies, baseJson.devDependencies);
        expect(mockMergeJsonObjects).not.toHaveBeenNthCalledWith(1, stubJson.dependencies, baseJson.dependencies);

        // check calls to writeJsonDataToFile
        expect(mockWriteJsonDataToFile).toHaveBeenNthCalledWith(1, BASE_FILE, updatedJsonNoBootstrap);

        // check the final snapshot
        expect(vol.toJSON()).toMatchSnapshot();
    });

    it("It should pass if the file package.json has been updated with the merged dependencies", async () => {

        vol.fromJSON(
            {
                [BASE_FILE]: JSON.stringify(baseJson),
                [BASE_STUB_FILE]: JSON.stringify(stubJson)
            },
            '/'
        );

        await addDependenciesToPackageJson(BASE_STUB_FILE, BASE_FILE, true);

        // check calls to getJsonDataFromFile
        expect(mockGetJsonDataFromFile).toHaveBeenNthCalledWith(1, BASE_STUB_FILE);
        expect(mockGetJsonDataFromFile).toHaveBeenNthCalledWith(2, BASE_FILE);

        // check calls to mergeJsonObjects
        expect(mockMergeJsonObjects).toHaveBeenNthCalledWith(1, stubJson.devDependencies, baseJson.devDependencies);
        expect(mockMergeJsonObjects).toHaveBeenNthCalledWith(2, stubJson.dependencies, baseJson.dependencies);

        // check calls to writeJsonDataToFile
        expect(mockWriteJsonDataToFile).toHaveBeenNthCalledWith(1, BASE_FILE, updatedJsonBootstrap);

        // check the final snapshot
        expect(vol.toJSON()).toMatchSnapshot();
    });
});