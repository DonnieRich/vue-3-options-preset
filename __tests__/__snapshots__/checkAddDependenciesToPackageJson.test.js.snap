// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addDependenciesToPackageJson Should pass if the file package.json has been updated with the merged dependencies 1`] = `
{
  "/node_modules/vue-3-options-preset/src/stubs/package.json": "{"dependencies":{"@popperjs/core":"^2.11.6","bootstrap":"^5.2.3"},"devDependencies":{"sass":"^1.57.1"}}",
  "/package.json": "{"license":"MIT","dependencies":{"shelljs":"^0.8.5","@popperjs/core":"^2.11.6","bootstrap":"^5.2.3"},"devDependencies":{"jest":"^29.3.1","memfs":"^3.4.12","sass":"^1.57.1"}}",
}
`;

exports[`addDependenciesToPackageJson Should pass if the file package.json has been updated with the merged devDependencies 1`] = `
{
  "/node_modules/vue-3-options-preset/src/stubs/package.json": "{"dependencies":{"@popperjs/core":"^2.11.6","bootstrap":"^5.2.3"},"devDependencies":{"sass":"^1.57.1"}}",
  "/package.json": "{"license":"MIT","dependencies":{"shelljs":"^0.8.5"},"devDependencies":{"jest":"^29.3.1","memfs":"^3.4.12","sass":"^1.57.1"}}",
}
`;
